### ISSUE TITLE
How to treat an unbalanced DGL Dataset?

### ISSUE THREAD

**Issue by Sinasta** [wassimj/topologicpy#5]
Is there an API included in _topologicpy_ to balance out an unbalanced DGL dataset?
I can't find anything geared towards that in the API documentation and i would like to ask you, what your plans/opinion is about including an API for that purpose.
To me it seems that there are two possible ways:

1.  The first one would be to include [`torch.utils.data.WeightedRandomSampler()`](https://pytorch.org/docs/stable/data.html#torch.utils.data.WeightedRandomSampler) in the `DGL.Train()` function (But this could be a bit more complicated because of the combination with the [`SubsetRandomSampler()`](https://discuss.pytorch.org/t/dataloader-using-subsetrandomsampler-and-weightedrandomsampler-at-the-same-time/29907))

2. The second option would be (and i think this one is more in line with the modular approach of _topologicpy_) to Include a function `DGL.BalanceDGLDataset` which would take a DGLDataset and the output dict of `DGL.CategoryDistribution()` as inputs and returns a new balanced DGL Dataset. A possible argument for this new function, could be the method of balancing (undersampling, oversampling et.).

If you are interested and if you see a point in implementing those functionalities i would be happy to try implementing one of those options. If not feel free to close this Issue

**Comment by wassimj** [comment]
Excellent suggestion. I prefer option 2. But I am unsure how oversampling would work. Would you just duplicate graphs of that category?
Another feature I wanted to add is the ability to remove all graphs of a certain category.
Yes, please contribute these methods if you have the time and inclination. Thank you!!

**Comment by Sinasta** [comment]
I think even for option 2. there are multiple options:
1. implementing a super simple undersampling (and oversampling through random duplication) algorithm with numpy's `np.random.choice` or pandas. The advantage would be that there is no need for an extra library but some of the random samples could be too similar and misrepresents the data set.
2. using a library like [imbalanced-learn](https://imbalanced-learn.org/stable/under_sampling.html) which would allow for some more complex over- and undersampling methods but would require additional libraries to be installed which i don't think is a good idea.
3. experimenting with an implementation of pytorch's WeightedRandomSampler to load a dataset with weights and then to return the balanced dataset (i honestly don't know if this is even conceptually feasible)

I have to admit that my knowledge in ML is extremely limited but if you tell me which of those options (or maybe you have another idea) makes the most sense to you then i would be happy to try experimenting.

**Comment by wassimj** [comment]
I think random sampling is good enough for now. Many thanks!

**Comment by wassimj** [comment]
This has been incorporated in release version 0.2.4. Please check and report back here so I can close this issue. Thank you.

**Comment by Sinasta** [comment]
It works perfectly, but I still think this should be used with caution, as these simple balancing methods (especially oversampling) risk changing the dataset in a non-optimal way
